src/index.jst

```diff
@@ -2,44 +2,45 @@
 
 // do not edit .js files directly - edit src/index.jst
 
+{{? it.es6 }}
+var envHasBigInt64Array = typeof BigInt64Array !== 'undefined';
+{{?}}
+
 module.exports = function equal(a, b) {
   if (a === b) return true;
 
-  if (a && b && typeof a == 'object' && typeof b == 'object') {
+  if (a && b && typeof a ==='object' && typeof b ==='object') {
     if (a.constructor !== b.constructor) return false;
 
-    var length, i, key, keys;
+    var length, i, key;
     if (Array.isArray(a)) {
       length = a.length;
-      if (length != b.length) return false;
+      if (length !== b.length) return false;
       for (i = length; i-- !== 0;)
         if (!equal(a[i], b[i])) return false;
       return true;
     }
 
 {{? it.es6 }}
+    var keys;
     if (a instanceof Map) {
       if (a.size !== b.size) return false;
-
-      keys = getKeys(a);
-      for (i = a.size; i-- !== 0;)
-        if (!b.has(keys[i])) return false;
-
-      for (i = a.size; i-- !== 0;) {
-        key = keys[i];
-        if (!equal(a.get(key), b.get(key))) return false;
+      // should be named `iterator` for the sake of variables efficiency
+      keys = a.entries();
+      while ((i = keys.next()) && !i.done) {
+        if (!b.has(i.value[0])) return false;
+        if (!equal(i.value[1], b.get(i.value[0]))) return false;
```

Comment:
Here are two things that improved performance:
- Omitted `getKeys` call, which decreases Map iterations by two times;
- Because we are equating values in the same loop - current item's value already available, so no need in extra `a.get()` call;

Reply:
I think the idea was that comparing the keys is faster than comparing the values, as the values can be deep.

Also is ES6 compatibility really a concern for the function that adds support for ES6 objects? The original equal is still ES5 compatible.

Reply:
> I think the idea was that comparing the keys is faster than comparing the values, as the values can be deep.

Ideally it would be faster for situations when it fast-fails with missing one of the first items in second Map. But it really slows things up when objects are equal.
And actually i just realised that my aproach can really slow everything down in case of negative result around the last values =\.


> Also is ES6 compatibility really a concern for the function that adds support for ES6 objects? The original equal is still ES5 compatible.
This one it gully ES5 compatible even for ES6 version (in terms of syntax).